### PR TITLE
Change implementation of texture coordinates on corners

### PR DESCRIPTION
**Pull request by MatthieuZins**
This PR adds a new texture coordinates mode: one coordinate per face corner.

### CODE REVIEW

File: vital/types/mesh.cxx [hunk 1]
```diff
@@ -419,6 +419,15 @@ ::merge(const mesh& other)
     {
       tex = std::vector<vector_2d>(2*num_e, vector_2d(0,0));
     }
+    else if (other.has_tex_coords() == TEX_COORD_ON_FACE_CORNER)
+    {
+      int nb_face_corners = 0;
+      for (int f=0; f < this->num_faces(); ++f)
```
Comment (dstoup) [inline vital/types/mesh.cxx]: Very minor nit. Put spaces between operators and operands. Same comment applies for all cases below.

**Comment by mleotta** [comment]
I'm missing some conceptual understanding here.  How is a "face corner" different from a "corner"?  The difference between vertices and corners is that multiple corners of faces will share a common vertex in the mesh, but you may need separate texture coordinates per corner if your texture space does not have the same mesh topology.

Why can you not use the existing corners (other than there is a missing implementation of finding which faces are valid)?

**Comment by MatthieuZins** [comment]
I agree that they are similar, but the number of "corners" is defined as two times the number of edges. It's ok when the mesh is watertight because each edge is adjacent to 2 faces. But it doesn't correspond to the number of "face corners" when the mesh has boundary edges.

**Comment by mleotta** [comment]
I would argue that defining the number of corners as two time the number of edges is bug then and should be corrected rather than defining a new type of texture coordinates to work around it.  I don't see any valid application that would need to distinguish between "corner" and "face corner".  It may still be helpful over estimate the number of corners to use edge indexing to quickly navigate between face corners, but then just set the invalid corners to (-1, -1) or some other invalid state and not write them out to files.  However, I haven't looked at this code in detail in a while, so I'm not certain.

**Comment by MatthieuZins** [comment]
Ok, then I see two options:

1) Define the number of corners as the number of "valid" half edges (i.e. the number of half edges with a valid adjacent face), or directly as "face corners" similarly to the new TEX_COORD_ON_FACE_CORNER that I added.

2) Keep the over-estimated number of corners and set invalid corners to (-1, -1).


On the one hand, the first solution does not permit indexing with half-edges, but if the mesh has regular faces it is possible to quickly navigate through texture coordinates using the face index. Also, if we disconnect corners from half-edges, we don't need to build the edge graph.

On the other hand, the second solution requires more steps when we input/output texture coordinates because they are generally indexed like this:

          face 0           face 1          face 2   
     vt00 vt01 vt02    vt10 vt11 vt12    vt20 vt21 vt22  ...

We would have to find the corresponding half-edge of each texture coordinate, rearrange the vector and add/remove the invalid values.

Do you have a preference?

**Comment by mleotta** [comment]
Maybe a hybrid.  I think approach number 1 works best when you have a regular mesh (e.g. all triangles or all quads).  That is the common case, we should probably handle the way you have proposed in this branch.  However, for irregular meshes, it would be more efficient to use half edge indexes for texture coordinates.  We could do it both ways depending on whether you have regular mesh or not.  This is just an implementation detail.  Either way the number of texture coordinates written to a file should be the same, which is why I don't think it should have a new enum value.

**Comment by mleotta** [comment]
Actually, I think it would be good to talk this over.  It just occurred to me that there is a case for a third type texture coordinates, so I'd like to discuss this.

**Comment by MatthieuZins** [comment]
Sounds good, tell me when you have some time next week.

**Comment by MatthieuZins** [comment]
Based on our discussion, I replaced TEX_COORD_ON_CORNER with my implementation.

**Comment by mleotta** [comment]
Jenkins test this please.

**Comment by Cookt2** [comment]
Jenkins test this please

**Comment by Cookt2** [comment]
Note that the exact string must be "Jenkins test this please". I'm going to add a few other variations to the trigger; to make it a bit more user-friendly.

**Pull request merged**